### PR TITLE
fix: traincascade, use C++ persistence API to compile with 4.x

### PR DESCRIPTION
**Pull request by fdivitto**
This pull allows to compile traincascade application with OpenCV 4.6. Changes uses new persistence C++ API in place of legacy one.

#13231

### CODE REVIEW

File: apps/traincascade/old_ml_tree.cpp [hunk 1]
```diff
@@ -3798,59 +3807,59 @@ void CvDTree::write_split( CvFileStorage* fs, CvDTreeSplit* split ) const
...
+    fs.write( "node_risk", node->node_risk );
+    fs.write( "tree_risk", node->tree_risk );
+    fs.write( "tree_error", node->tree_error );
 
     if( node->left )
     {
-        cvStartWriteStruct( fs, "splits", CV_NODE_SEQ );
+        fs.startWriteStruct( "splits", FileNode::SEQ );
 
         for( split = node->split; split != 0; split = split->next )
```
Comment (asmorkalov) [inline apps/traincascade/old_ml_tree.cpp:3852]: `for( CvDTreeSplit* split = node->split; split != 0; split = split->next )` and remove `CvDTreeSplit* split;` definition.

**Comment by asmorkalov** [comment]
Also please pay attention on CI failures.

File: apps/traincascade/old_ml_inner_functions.cpp [hunk 2]
```diff
@@ -60,61 +60,57 @@ void CvStatModel::clear()
...
-    CV_CALL( fs = cvOpenFileStorage( filename, 0, CV_STORAGE_WRITE ));
-    if( !fs )
+    if( !fs.open( filename, cv::FileStorage::WRITE ))
         CV_ERROR( CV_StsError, "Could not open the file storage. Check the path and permissions" );
 
     write( fs, name ? name : default_model_name );
 
     __END__;
 
-    cvReleaseFileStorage( &fs );
```
[the remark this answers is not in the document]
Comment (fdivitto) [reply apps/traincascade/old_ml_inner_functions.cpp:77]: fs destructor should call release()

File: apps/traincascade/old_ml_boost.cpp [hunk 3]
```diff
@@ -1899,35 +1899,35 @@ void CvBoost::write_params( CvFileStorage* fs ) const
...
 
-void CvBoost::read_params( CvFileStorage* fs, CvFileNode* fnode )
+void CvBoost::read_params( cv::FileNode& fnode )
 {
     CV_FUNCNAME( "CvBoost::read_params" );
 
     __BEGIN__;
 
-    CvFileNode* temp;
+    cv::FileNode temp;
```
[the remark this answers is not in the document]
Comment (fdivitto) [reply apps/traincascade/old_ml_boost.cpp:1924]: ok

File: apps/traincascade/old_ml_boost.cpp [hunk 4]
```diff
@@ -1937,71 +1937,71 @@ void CvBoost::read_params( CvFileStorage* fs, CvFileNode* fnode )
...
 
-    if( temp && CV_NODE_IS_STRING(temp->tag) )
+    if ( temp.isString() )
     {
-        const char* boost_type_str = cvReadString( temp, "" );
-        params.boost_type = strcmp( boost_type_str, "DiscreteAdaboost" ) == 0 ? DISCRETE :
-                            strcmp( boost_type_str, "RealAdaboost" ) == 0 ? REAL :
-                            strcmp( boost_type_str, "LogitBoost" ) == 0 ? LOGIT :
-                            strcmp( boost_type_str, "GentleAdaboost" ) == 0 ? GENTLE : -1;
+      auto boost_type_str = (std::string) temp;
```
[the remark this answers is not in the document]
Comment (fdivitto) [reply apps/traincascade/old_ml_boost.cpp:1946]: ok

File: apps/traincascade/old_ml_tree.cpp [hunk 5]
```diff
@@ -4108,34 +4117,34 @@ void CvDTree::read_tree_nodes( CvFileStorage* fs, CvFileNode* fnode )
 }
 
 
-void CvDTree::read( CvFileStorage* fs, CvFileNode* fnode )
+void CvDTree::read( cv::FileNode fnode )
```
[the remark this answers is not in the document]
Comment (fdivitto) [reply apps/traincascade/old_ml_tree.cpp:4120]: ok

File: apps/traincascade/old_ml_tree.cpp [hunk 6]
```diff
@@ -3884,43 +3893,43 @@ void CvDTree::write_tree_nodes( CvFileStorage* fs ) const
...
     write_tree_nodes( fs );
-    cvEndWriteStruct( fs );
+    fs.endWriteStruct();
 
     __END__;
 }
 
 
-CvDTreeSplit* CvDTree::read_split( CvFileStorage* fs, CvFileNode* fnode )
+CvDTreeSplit* CvDTree::read_split( cv::FileNode fnode )
```
[the remark this answers is not in the document]
Comment (fdivitto) [reply apps/traincascade/old_ml_tree.cpp:3932]: ok

File: apps/traincascade/old_ml_tree.cpp (hunk 1 above)
Comment (fdivitto) [reply apps/traincascade/old_ml_tree.cpp:3852]: ok

**Comment by fdivitto** [comment]
> Also please pay attention on CI failures.

fixed

**Comment by fdivitto** [comment]
Hello,
after I have fixed as your suggestions, I have to make a new commit and push?

Best regards,
Fabrizio Di Vittorio

**Comment by asmorkalov** [comment]
Yes.

**Pull request closed without merging**